content/authoring/kata/floating-point.md

```diff
@@ -0,0 +1,277 @@
+
+
+#### Rounding does not help
+
+Sometimes authors try to work around the problems with floating-point comparisons in some strange ways, like rounding, stringification, or even other, worse things. However such workarounds are a bad thing and they only make things worse. Concept of _"rounding to n decimal places"_ is not well defined for floating-point numbers, and it does not always work as expected. Rounding to an integer value, be it floor, ceil, round to nearest, or truncation, also is not guaranteed to work and often leads to errors. Do not do this, or you will only make things worse. The easiest way to get things right is to require no rounding, and use proper assertions with tolerance.
+
+
+### Be careful when formatting
+
+Some existing kata can issue such a very confusing assertion message:
+
+```c
+Test(sample_test, test_basic) {
+  double actual = squareArea(1.1);
+  double expected = 1.21;
+  cr_assert_float_eq(actual, expected, 1e-9, "Expected %f +/- 1e-9, but was %f", expected, actual);
+}
+```
+
+```text
+Test Results:
+sample_test
+    test_basic
+        Expected 1.210000 +/- 1e-9, but was 1.210000
+```
+
+This issue is not limited only to C, other languages are also affected. 
+
+So... what's wrong? Why tests fail, if expected value and actual value are both `1.210000`?
+Most probably, two things are wrong:
+- User solution returns invalid answer, which does not fall into required tolerance, and
+- Tests display assertion message incorrectly, truncating some meaningful digits.
+
+Actually, both numbers are different, and actual answer is wrong by more than allowed tolerance. But when formatting floating-point values, many languages use only some part of digits, and truncate the least meaningful ones, even when they are still important. When you fix formatting of the message, you get better feedback:
+
+```c
+//note %.9f as format specifiers
+cr_assert_float_eq(actual, expected, 1e-9, "Expected %.9f +/- 1e-9, but was %.9f", expected, actual);
+```
+
+```text
+Test Results:
+sample_test
+    test_basic
+        Expected 1.210000000 +/- 1e-9, but was 1.210000343
+```
+
+You need to remember that whenever you print or format floating-point values as strings, they should always be formatted with at least as many significant digits as the tolerance used for comparisons.
+
+
+### Use alternatives
+
+There are some ways to mitigate some issues related to floating-point numbers by using some other means, language constructs, classes, libraries, etc. Unfortunately, they are usually language-specific and can turn out to be problematic in kata, because translating kata between languages can become difficult. However, some possibilities are:
+
+- You can use arbitrary-precision decimal types or similar, for example `java.math.BigDecimal` in Java, `decimal` (floating-point decimal primitive type) in C#, `decimal` module in Python, etc.
```

Comment:
> some ways to mitigate some issues related to floating-point numbers

Using floating point numbers to mitigate issues with floating point numbers? Arbitrary precision is still finite, so it's just binary vs decimal here, which makes the difference only if the input is a finite decimal.

Reply:
I was wondering how/if to differentiate between decimal fixed-scale numbers (`java.math.BigDecimal`), floating-point decimal numbers, (`decimal` in C#), all of them as opposed to floating-point binary numbers (IEEE-754). So basically the sentence would mean something like _"Using decimal floating point numbers to mitigate issues with IEEE-754 floating point numbers"_ . Or something like that.

I am not sure about the _" it's just binary vs decimal"_, or maybe I am not sure about a proper nomenclature, because some nitpicker could argue that C# `decimal` is still binary (IIRC it's encoded in some form of BCD).

But yeah, maybe just "use decimal data type if appropriate" would be enough, I am just afraid that someone could get an impression that decimal data types from different libraries/languages are functionally equivalent.

Reply:
I've reworded this part a bit, got rid of unnecessary deails.